tools/fs/watch.js

```diff
@@ -265,6 +266,93 @@ export function sha1(...args) {
+//
+// Known issues: 
+// - single character ? and range of character [a-z] are not supported. (not in the git man page)
+// - Update rules for subdirectories in .meteorignore (like client/*.ts in the app directory) 
+//   will not activate autoupdate. You need to restart meteor for the case.
+//   If you want autoupdated .meteorignore for a directory 
+//   you need to write the directory's own .meteorignore in the directory.
+// 
+// Written by: Isaac Han
+// Github id: i4han
+//
+
+const METEOR_IGNORE = '.meteorignore'
+
+var appDir = files.findAppDir(files.cwd());
+var mtimeMeteorIgnore = {};
+var cacheMeteorIgnore = {};
+
+function readMeteorIgnore(path) {
+  var fileMeteorIgnore = files.pathJoin(path, METEOR_IGNORE)
+  var stat = files.statOrNull(fileMeteorIgnore)
+  if (stat === null) { // .meteorignore is not found.
+    return [];  
+  } else if (
+    // stat.mtime is an object. It needs to be converted to sting to compare
+    mtimeMeteorIgnore[path] === stat.mtime.toString() && 
+    cacheMeteorIgnore.hasOwnProperty(path) 
+  ) {
+    return cacheMeteorIgnore[path];   // use the cached instead of the file.
+  } else {
+    mtimeMeteorIgnore[path] = stat.mtime.toString(); // store the modified time to check the cache.
+    return cacheMeteorIgnore[path] = files.readFile(fileMeteorIgnore)
+      .toString('utf8')
+      .split(/\n/)
+      .filter(v => v);  // remove '' line. Don't trim because there may be \(space) at the end of line.
+  }
+}
+
+function filterMeteorIgnore(path) {
+  var rules = readMeteorIgnore(path)
+  // move up path .split('/').slice(0, -1).join('/') until it arrives to addDir or ''
+  while (appDir !== path && path) {
+    path = path.split('/').slice(0, -1).join('/');
+    rules = rules.concat(readMeteorIgnore(path));  
+  }
+  return rules.length
+    ? ignore().add(rules)
+    : null;
+}
+
+function isInPath(source, str) {
+  return source.indexOf('/' + str + '/') > -1 || source.endsWith('/' + str)
+}
+
+
+function meteorIgnore(absPath, contents) {
+  if (
+    !isInPath(absPath, 'node_modules') && 
+    !isInPath(absPath, 'imports') &&            // skip if current path is in node_modules or imports
+    absPath.slice(0, appDir.length) === appDir  // check if current path is in the appDir
+  ) { 
+    var ig = filterMeteorIgnore(absPath)
+    if (ig) {
+      contents = ig.filter(           // contents are a list of basenames (no path just filename). 
+        contents.map(v =>             // In order to be used by ignore filter
+          files.pathJoin(absPath, v)  // path from the *appDir* should be added to the contents list.  
+            .slice(appDir.length + 1) // A full path (not from appDir) is given, it should be cut like 'client/'
+        )                             // Otherwise it will not be filtered by 'client/' in .meteorignore.
+      ).map(v => v.split('/').slice(-1)[0]); // This line is to make path.basename. Back to list of basename.
```

Comment:
Can you keep lines wrapped at 80 chars? Also - comments can be helpful, but I think in this case an end of line comment on each line is a bit excessive and detracts from code readability. Could you maybe summarize these comments a bit more and move them above the `filter` / `map` calls?